edb/server/tenant.py

```diff
@@ -965,11 +1025,26 @@ async def acquire_pgcon(self, dbname: str) -> pgcon.PGConnection:
 
         for _ in range(self._pg_pool.max_capacity):
             conn = await self._pg_pool.acquire(dbname)
-            if conn.is_healthy():
-                return conn
+            if not conn.is_healthy():
+                logger.warning("acquired an unhealthy pgcon; discard now")
+            elif conn.last_init_con_data is not self._init_con_data:
+                if self._update_init_con_sql:
+                    try:
+                        await conn.sql_execute(self._update_init_con_sql)
```

Comment:
Is there any way to avoid doing this every time?  Ideally we should track whether a connection has the updated state.  Maybe a dict with version counter or something like that?

Reply:
`last_init_con_data` would be for "track whether a connection has the updated state"? The `is` reference check should be as fast as a version counter.

Reply:
I'm a bit confused about the interaction between `_update_init_con_sql` and that check then. Can there be a case when the state is not matching but `_update_init_con_sql` is `None`?

Reply:
Theoretically no - whenever we update the state, we set a new update SQL. Let me change the code a bit to ensure that it is always done "atomically" with a single function.

In the worst case when it's really happening, it prints a warning, discard that pgcon and acquire a new one, and the new one may just `pg_connect()` with a proper state.